### DISCUSSION TITLE
whereNotIn and when param class type is not same

### DISCUSSION THREAD

**Discussion by mowangjuanzi** [laravel/framework#35810] [Q&A] [answered]
```php
User::whereNotIn("id", function ($builder1) {
        
    })->when(1, function (QueryBuilder $builder2) {
        
    }))
```

the class type of variable `$builder1` is `Illuminate\Database\Query\Builder`

the class type of variable `$builder2` is `Illuminate\Database\ Eloquent\Builde`

Why can't variables `$builder1` and `$builder2` be of the same type?

**Comment by veelasky** [comment, marked as answer]
because all query builder that operate on the database, i.e : `where`, `select`, etc, is directly operate on the PDO, if you ever use basic query builder using `DB` facade, that's why the signature is basic query builder.

now all "helpers" method such as `when`, `scope` is directly connected with eloquent, thus. eloquent builder is used.

note: since eloquent builder is also instance of basic query builder, all basic query method also implemented.

**Reply by mowangjuanzi** [reply]
thank you
